spec/emerald_spec.rb

```diff
@@ -0,0 +1,129 @@
+
+        lisp_file_contents = File.read(test_file)
+        compiler = Emerald::Compiler.new(lisp_file_contents)
+
+        expect{ compiler.compile }.to raise_error(Emerald::Parser::InvalidListError).with_message(Emerald::Parser::ERROR_INVALID_LIST)
+      end
+    end
+
+    context "when it receives an invalid list without an opening bracket" do
+      it "raises an InvalidListError" do
+        test_file.write('+ 5 2 )')
+        test_file.close
+
+        allow(user_comms).to receive(:new).with(stdout)
+        allow(user_comms).to receive(:verify_lisp_file).and_return(test_file)
+        lisp_file_contents = File.read(test_file)
+        compiler = Emerald::Compiler.new(lisp_file_contents)
+
+        expect{ compiler.compile }.to raise_error(Emerald::Parser::InvalidListError).with_message(Emerald::Parser::ERROR_INVALID_LIST)
+      end
+    end
+
+
+    context "when it receives an invalid lisp function expression" do
+      it "raises an InvalidListError" do
+        test_file.write('( 5 + 5 )')
+        test_file.close
+
+        allow(user_comms).to receive(:new).with(stdout)
+        allow(user_comms).to receive(:verify_lisp_file).and_return(test_file)
+        lisp_file_contents = File.read(test_file)
+        compiler = Emerald::Compiler.new(lisp_file_contents)
+
+        expect{ compiler.compile }.to raise_error(Emerald::Rubify::InvalidLispFunctionError)
+      end
+    end
+
+
+    context "when it receives a valid lisp expression with a single number" do
+      it "compiles and evaluates the lisp" do
+        test_file.write('5')
+        test_file.close
+        stdout = `ruby lib/emerald.rb "#{test_file.path}"`.chomp
+        expect(stdout).to eq('5.0')
+      end
+    end
+
+    context "when it receives a valid lisp expression with an addition operation" do
+      it "compiles and evaluates the lisp" do
+        test_file.write('(+ 2 5)')
+        test_file.close
+
+        stdout = `ruby lib/emerald.rb "#{test_file.path}"`.chomp
+        expect(stdout).to eq('7.0')
+      end
+    end
+
+    context "when it receives a valid lisp expression with a subtraction operation" do
+      it "compiles and evaluates the lisp" do
+        test_file.write('(- 5 2 1)')
+        test_file.close
+
+        stdout = `ruby lib/emerald.rb "#{test_file.path}"`.chomp
+        expect(stdout).to eq('2.0')
+      end
+    end
+
+    context "when it receives a valid lisp expression with a division operation" do
+      it "compiles and evaluates the lisp" do
+        test_file.write('(/ 20 2 2)')
+        test_file.close
+
+        stdout = `ruby lib/emerald.rb "#{test_file.path}"`.chomp
+        expect(stdout).to eq('5.0')
+      end
+    end
+
+    context "when it receives a valid lisp expression with a multiplication operation" do
+      it "compiles and evaluates the lisp" do
+        test_file.write('(* 12 2 2)')
+        test_file.close
+
+        stdout = `ruby lib/emerald.rb "#{test_file.path}"`.chomp
+        expect(stdout).to eq('48.0')
```

Comment:
Maybe not quite the right place to ask but how some we use floats internally when the source code has what looks like integers?